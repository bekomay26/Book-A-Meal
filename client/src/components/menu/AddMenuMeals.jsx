import React from 'react';
import PropTypes from 'prop-types';


const AddMenuMeals = ({ onSave, onSelect, meals }) => (
  <div>
    <form>
      <div className="row">
        <div className="col-12 subm">
          <input type="submit" value="Save Menu" onClick={onSave} />
        </div>
      </div>
      <div className="menu-items">
        <ul id="todayslist">
          {
            meals.map(meal => (
              <li
                key={meal.id}
                onKeyPress={event => onSelect(event, meal)}
                onClick={event => onSelect(event, meal)}
                tabIndex="0"
                role="button"
              >
                {meal.title}<span className="sel-menu-item-price">&#x20A6; {meal.price}</span>
              </li>))
          }
        </ul>
      </div>
    </form>
  </div>
);

AddMenuMeals.propTypes = {
  onSave: PropTypes.func.isRequired,
  onSelect: PropTypes.func.isRequired,
  meals: PropTypes.array.isRequired,
};

export default AddMenuMeals;
