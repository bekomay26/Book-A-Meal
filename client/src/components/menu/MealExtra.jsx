import React from 'react';
import PropTypes from 'prop-types';
import { Checkbox, Grid } from 'semantic-ui-react';

const MealExtra = ({ extra, onQtyChange, extraStatus, handleChecked, indexKey }) => {
  const extStatPosition = extraStatus.findIndex(extraStat => extraStat.key === extra.id);
  let extraQty = 1;
  if (extraStatus[extStatPosition] !== undefined) {
    extraQty = extraStatus[extStatPosition].qty;
  } else {
    extraQty = 1;
  }

  return (<div>
    <Grid divided="vertically">
      <Grid.Row>
        <Grid.Column width={8}>
          <Checkbox
            onChange={(event) => handleChecked(event, extra.id, extra, extraQty)}
            className="menu-modalextra-text"
            label={extra.title}
          />
        </Grid.Column>
        <Grid.Column width={4} className="menu-modalextra-qty-grid">
          <input
            type="number"
            className="col-3 menu-modalextra-qty"
            onInput={(event) => onQtyChange(event, extra.id, extra.price)}
            defaultValue={1}
            min={1}
          />
        </Grid.Column>
        <Grid.Column width={4}>
          <div className="menu-modalextra-text">
            &#x20A6;{extra.price * extraQty}
          </div>
        </Grid.Column>
      </Grid.Row>
    </Grid>
  </div>)
};

MealExtra.propTypes = {
  extra: PropTypes.shape({
    title: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
  })
    .isRequired,
};

export default MealExtra;
