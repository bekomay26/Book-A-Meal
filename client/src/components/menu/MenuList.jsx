import React from 'react';
import PropTypes from 'prop-types';
import MenuItem from './MenuItem';


const MenuList = ({ meals, selected }) => (
  <div>
    {meals.map(meal =>
      <MenuItem meal={meal} selected={selected} />)
    }
  </div>
);

MenuList.propTypes = {
  meals: PropTypes.arrayOf(PropTypes.shape).isRequired,
  selected: PropTypes.func.isRequired,
};

export default MenuList;
