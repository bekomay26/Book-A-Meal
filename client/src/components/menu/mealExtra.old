import React from 'react';
import CSSModules from 'react-css-modules';
import PropTypes from 'prop-types';
import styles from '../../assets/styles/menu2.css';
import { Checkbox } from 'semantic-ui-react';

const MealExtra = ({ extra, onQtyChange, extraStatus, handleChecked, indexKey }) => {
  let extraQty = 1;
  if (extraStatus[indexKey] !== undefined) {
    extraQty = extraStatus[indexKey].qty;
  } else {
    extraQty = 1;
  }
  return (<div>
    <div className="row">
      <div className="col-6">
        <input type="checkbox" name="" value="" onChange={(event) => handleChecked(event, indexKey, extra)} /><span>{extra.title}</span>
      </div>
      <div className="col-3">
        <input type="number" className="col-3" onInput={(event) => onQtyChange(event, indexKey)} defaultValue={1} />
      </div>
      <span className="col-3 foodprice">&#x20A6;{extra.price * extraQty}</span>
    </div>
  </div>)
};

MealExtra.propTypes = {
  extra: PropTypes.shape({
    title: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
  })
    .isRequired,
};

// export default CSSModules(MealExtra, styles, { allowMultiple: true });
export default MealExtra;
