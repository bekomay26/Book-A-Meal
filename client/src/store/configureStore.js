import { createStore, applyMiddleware } from 'redux';
import reduxImmutableStateInvariant from 'redux-immutable-state-invariant';
import thunk from 'redux-thunk';
import rootReducer from '../reducers/index';


const configureStore = initialState =>
  createStore(
    rootReducer,
    initialState,
    applyMiddleware(thunk, reduxImmutableStateInvariant()),
  );

export default configureStore;
