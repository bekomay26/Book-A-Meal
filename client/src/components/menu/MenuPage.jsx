import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Drawer } from 'antd';
import CSSModules from 'react-css-modules';
import { loadMenu } from '../../actions/menuActions';
import { saveOrder } from '../../actions/orderActions';
import { logout } from '../../actions/authActions';
import UserNavBar from '../common/UserNavBar';
import styles from '../../assets/styles/menu2.css';
import MenuList from './MenuList';
import MenuForm from './MenuForm';

export class MenuPage extends Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      // meal: Object.assign({}, this.props.meal),
      meal: {},
      extraQty: 1,
      extraStatus: [],
      selectedExtras: [],
      saving: false,
      basePrice: 0,
      totalPrice: 0,
      visible: false,
      isDesktop: false,
      visibleSideBar: false,
    };
    // this.onSelect = this.onSelect.bind(this);
    this.onSave = this.onSave.bind(this);
    this.onClose = this.onClose.bind(this);
    this.myMeal = this.myMeal.bind(this);
    this.goesWith = this.goesWith.bind(this);
    this.onTop = this.onTop.bind(this);
    this.onQtyChange = this.onQtyChange.bind(this);
    this.handleChecked = this.handleChecked.bind(this);
    this.showDrawer = this.showDrawer.bind(this);
    this.onCloseDrawer = this.onCloseDrawer.bind(this);
    this.updatePredicate = this.updatePredicate.bind(this);
  }

  componentDidMount() {
    this.props.loadMenu();// just added
    this.updatePredicate();
    window.addEventListener('resize', this.updatePredicate);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.updatePredicate);
  }

  updatePredicate() {
    this.setState({
      isDesktop: window.innerWidth > 960
    });
  }

  async myMeal (currentMeal) {
    this.showDrawer();
    // this.setState({meal:this.props.meal});
    this.setState({ meal: currentMeal, basePrice: currentMeal.price, totalPrice: currentMeal.price });

    const newExtraStatus = [];
    // edit extra arrangement wrong
    currentMeal.extras.forEach((ext) => newExtraStatus.push({ isChecked: false, qty: 1, extra: ext, key: ext.id, price: ext.price } ));
    await this.setState({ extraStatus: newExtraStatus });
  }

  
  // onOpen() {
  //   console.log(this.state.meal.price);
  //   this.setState({ totalPrice: this.state.meal.price });
  // }

  async onSave(e) {
    e.preventDefault();
    const extrasInfo = this.state.extraStatus.filter(extra => extra.isChecked === true);
    let extraIds = [];
    let extraQtys = [];
    extrasInfo.forEach(ext => extraIds.push(ext.extra.id));
    extrasInfo.forEach(ext => extraQtys.push(ext.qty));
    this.setState({ saving: true });
    console.log('bgbfere');
    await this.props.saveOrder({
      mealId: this.state.meal.id,
      extraIds,
      qtys: extraQtys,
      address: 'dsffsfgfsfvfsvfjysvfbgctf',
    });
    this.onCloseDrawer();
    this.redirect();
  }

  redirect() {
    this.setState({ saving: false });
  }

  // onSelect() {
  //   document.getElementById("modal-container").style.display ="block";
  // }

  onClose() {
    this.clearExtrasFields();
    document.getElementById('modal-container').style.display = 'none';
  }
  goesWith(mealExtras) {
    if (mealExtras !== undefined) {
      // console.log(mealExtras);
      return mealExtras.filter(extra => extra.category === 'GoesWith');
    } else
      return []
  }
  onTop(mealExtras) {
    if (mealExtras !== undefined)
      return mealExtras.filter(extra => extra.category === 'OnTop');
    else
      return []
  }

  clearExtrasFields() {
    this.setState({
      extraStatus: [],
      meal: {},
    });
  }

  onQtyChange(event, key, extraPrice) {
    const status = this.state.extraStatus;
    const extStatPosition = status.findIndex(extraStat => extraStat.key === key);
    status[extStatPosition].qty = parseInt(event.target.value, 10);
    status[extStatPosition].price = extraPrice * (status[extStatPosition].qty);
    let totalMealPrice = this.state.totalPrice;
    let mealPrice = this.state.basePrice;
    for (let i = 0; i < status.length; i += 1) {
      if (status[i].isChecked === true) {
        mealPrice += status[i].price;
      }
    }
    totalMealPrice = mealPrice;
    this.setState({ extraStatus: status, totalPrice: totalMealPrice });
  }

  handleChecked(event, key, extra, extraQuantity) {
    const checked = this.state.extraStatus;
    const selectedExt = this.state.selectedExtras;
    const extStatPosition = checked.findIndex(extraStat => extraStat.key === key);
    let totalMealPrice = this.state.totalPrice;
    if (checked[extStatPosition].isChecked === false) {
      selectedExt.push(extra);
      totalMealPrice += (extra.price * extraQuantity); // If checked, add extra price
    } else {
      const selectedExtPosition = selectedExt.findIndex(extraStat => extraStat.key === key);
      selectedExt.splice(selectedExtPosition, 1);
      totalMealPrice -= (extra.price * extraQuantity); // If unchecked, sub extra price
    }
    // checked[key].isChecked = !checked[key].isChecked;
    checked[extStatPosition].isChecked = !checked[extStatPosition].isChecked;
    this.setState({ extraStatus: checked, selectedExtras: selectedExt, totalPrice: totalMealPrice });
  }

  // onQtyChange(event, key, extraPrice) {
  //   // this.setState({ extraQty: parseInt(event.target.value, 10) });
  //   const status = this.state.extraStatus;
  //   status[key].qty = parseInt(event.target.value, 10);
  //   status[key].price = extraPrice * (status[key].qty);
  //   // this.setState({ extraStatus: status });
  //   let totalMealPrice = this.state.totalPrice;
  //   let mealPrice = this.state.basePrice;
  //   for (let i = 0; i < status.length; i += 1) {
  //     if (status[i].isChecked === true) {
  //       mealPrice += status[i].price;
  //     }
  //   }
  //   totalMealPrice = mealPrice;
  //   // if (status[key].isChecked === true) {
  //   //   totalMealPrice = this.state.basePrice + status[key].price;
  //   // }
  //   this.setState({ extraStatus: status, totalPrice: totalMealPrice });
  //   console.log(this.state.extraStatus);
  // }

  // handleChecked(event, key, extra, extraQuantity) {
  //   const checked = this.state.extraStatus;
  //   const selectedExt = this.state.selectedExtras;
  //   let totalMealPrice = this.state.totalPrice;
  //   if (checked[key].isChecked === false) {
  //     selectedExt.push(extra);
  //     totalMealPrice += (extra.price * extraQuantity); // If checked, add extra price
  //   } else {
  //     selectedExt.splice(key, 1);
  //     totalMealPrice -= (extra.price * extraQuantity); // If unchecked, sub extra price
  //   }
  //   checked[key].isChecked = !checked[key].isChecked;
  //   this.setState({ extraStatus: checked, selectedExtras: selectedExt, totalPrice: totalMealPrice });
  //   // console.log(this.state.extraStatus);
  //   // console.log(this.state.selectedExtras);
  //   console.log(totalMealPrice);
  // }

  showDrawer() {
    this.setState({
      visible: true,
    });
  }

  onCloseDrawer() {
    this.setState({
      visible: false,
    });
  }

  /**
   * @desc runs before component mounts
   * @memberof DocumentView
   * @returns {*} has no return value;
   */
  // componentWillMount() {
  //   this.props.loadMenu().then(() => {
  //     this.setState({
  //       menu: this.props.menu,
  //     });
  //   });
  // }
  render() {
    // const { menu } = this.state;
    const { menu } = this.props;
    const { userName } = this.props;
    const { isDesktop } = this.state;
    const { isAuthenticated } = this.props;

    // Work on
    // const goeswith = mealExtras => mealExtras.filter(extra => extra.category === 'GoesWith');
    // const extras = mealExtras => mealExtras.filter(extra => extra.category === 'OnTop');
    // const meaa = menu.pop();
    return (
      <div>
        <UserNavBar logout={this.props.logout} isAuthenticated={isAuthenticated} userName={userName} visibleSideBar />
        <div id="container" className="menu-page-container">
          {isDesktop ? (
            <h1>Today's Menu</h1>) : null
          }
          {/* <div>
          {menu}
          </div> */}

          <MenuList meals={menu} selected={this.myMeal} />
          <Drawer
            title="Order Meal"
            width={400}
            placement="right"
            onClose={this.onCloseDrawer}
            maskClosable={false}
            visible={this.state.visible}
            style={{
              height: 'calc(100% - 55px)',
              overflow: 'auto',
              paddingBottom: 53,
            }}
          >
            <MenuForm
              meal={this.state.meal}
              goesWith={this.goesWith}
              extras={this.onTop}
              onQtyChange={this.onQtyChange}
              extraStatus={this.state.extraStatus}
              handleChecked={this.handleChecked}
              totalMealPrice={this.state.totalPrice}
              onSave={this.onSave}
              onClose={this.onClose}
              saving={this.state.saving}
            />
          </Drawer>
        </div>
      </div>
    );
  }
}

MenuPage.defaultProps = {
  userName: undefined,
  isAuthenticated: false,
};

MenuPage.propTypes = {
  menu: PropTypes.arrayOf(PropTypes.shape).isRequired,
  // menu: PropTypes.shape({
  //   success: PropTypes.bool.isRequired,
  //   message: PropTypes.string.isRequired,
  //   meals: PropTypes.arrayOf(PropTypes.shape).isRequired,
  // })
  //   .isRequired,
  loadMenu: PropTypes.func.isRequired,
  saveOrder: PropTypes.func.isRequired,
  logout: PropTypes.func.isRequired,
  isAuthenticated: PropTypes.bool,
  userName: PropTypes.string,
  // meal: PropTypes.shape({
  //   id: PropTypes.number.isRequired,
  //   title: PropTypes.string.isRequired,
  //   description: PropTypes.string.isRequired,
  //   price: PropTypes.number.isRequired,
  //   extras: PropTypes.shape.isRequired,
  // })
  //   .isRequired,
};

/**
 * @desc maps dispatch to props;
 * @param {*} dispatch dispatch
 * @returns {*} action to be dispatched
 */
const mapDispatchToProps = dispatch => bindActionCreators({
  loadMenu,
  saveOrder,
  logout,
}, dispatch);

/**
 * @desc maps state to props;
 * @param {*} state store state
 * @returns {*} store state
 */
const mapStateToProps = state => ({
  menu: state.menuReducer,
  isAuthenticated: state.authReducer.isAuthenticated,
  userName: state.authReducer.userName,
});

const MenuPageWithCSS = CSSModules(MenuPage, styles);
export default connect(mapStateToProps, mapDispatchToProps)(MenuPageWithCSS);
