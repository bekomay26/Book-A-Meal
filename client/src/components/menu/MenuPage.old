import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Drawer } from 'antd';
import CSSModules from 'react-css-modules';
import { loadMenu } from '../../actions/menuActions';
import { saveOrder } from '../../actions/orderActions';
import { logout } from '../../actions/authActions';
import UserNavBar from '../common/UserNavBar';
import styles from '../../assets/styles/menu2.css';
import MenuList from './MenuList';
import MenuModal from './MenuModal';

class MenuPage extends Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      // meal: Object.assign({}, this.props.meal),
      meal: {},
      extraQty: 1,
      extraStatus: [],
      selectedExtras: [],
      saving: false,
      totalPrice: 0,
    };
    // this.onSelect = this.onSelect.bind(this);
    this.onSave = this.onSave.bind(this);
    this.onClose = this.onClose.bind(this);
    this.myMeal = this.myMeal.bind(this);
    this.goesWith = this.goesWith.bind(this);
    this.onTop = this.onTop.bind(this);
    this.onQtyChange = this.onQtyChange.bind(this);
    this.handleChecked = this.handleChecked.bind(this);
  }

  async myMeal (currentMeal) {
    document.getElementById("modal-container").style.display ="block";
    // this.setState({meal:this.props.meal});
    this.setState({ meal: currentMeal, totalPrice: currentMeal.price });
    console.log(`select ${currentMeal.title}`);

    const newExtraStatus = [];
    // edit extra arrangement wrong
    currentMeal.extras.forEach((ext) => newExtraStatus.push({ isChecked: false, qty: 1, extra: ext} ));
    await this.setState({ extraStatus: newExtraStatus });
    console.log(this.state.extraStatus);
    // return currentMeal;
  }

  onOpen() {
    console.log(this.state.meal.price);
    this.setState({ totalPrice: this.state.meal.price });
  }

  async onSave(e) {
    e.preventDefault();
    const extrasInfo = this.state.extraStatus.filter(extra => extra.isChecked === true);
    let extraIds = [];
    let extraQtys = [];
    extrasInfo.forEach(ext => extraIds.push(ext.extra.id));
    extrasInfo.forEach(ext => extraQtys.push(ext.qty));
    // console.log(this.state.meal.id);
    // console.log(extraIds);
    // console.log(extraQtys);
    this.setState({ saving: true });
    await this.props.saveOrder({
      mealId: this.state.meal.id,
      extraIds,
      qtys: extraQtys,
      address: 'dsffsfgfsfvfsvfjysvfbgctf',
    });
    this.onClose();
    this.redirect();
  }

  redirect() {
    this.setState({ saving: false });
    this.context.router.push('/cart');
  }

  // onSelect() {
  //   document.getElementById("modal-container").style.display ="block";
  // }

  onClose() {
    this.clearExtrasFields();
    document.getElementById("modal-container").style.display ="none";
  }
  goesWith(mealExtras) {
    if (mealExtras !== undefined) {
      
      return mealExtras.filter(extra => extra.category === 'GoesWith');
    } else
      return []
  }
  onTop(mealExtras) {
    if (mealExtras !== undefined)
      return mealExtras.filter(extra => extra.category === 'OnTop');
    else
      return []
  }

  clearExtrasFields() {
    this.setState({
      extraStatus: [],
      meal: {},
    });
  }

  onQtyChange(event, key) {
    // this.setState({ extraQty: parseInt(event.target.value, 10) });
    const status = this.state.extraStatus;
    status[key].qty = parseInt(event.target.value, 10);
    this.setState({ extraStatus: status });
    console.log(this.state.extraStatus);
  }

  handleChecked(event, key, extra, extraQuantity) {
    let checked = this.state.extraStatus;
    let selectedExt = this.state.selectedExtras;
    let totalMealPrice = this.state.totalPrice;
    if (checked[key].isChecked === false) {
      selectedExt.push(extra);
      totalMealPrice += (extra.price * extraQuantity); // If checked, add extra price
    } else {
      selectedExt.splice(key, 1);
      totalMealPrice -= (extra.price * extraQuantity); // If unchecked, sub extra price
    }
    checked[key].isChecked = !checked[key].isChecked;
    this.setState({ extraStatus: checked, selectedExtras: selectedExt, totalPrice: totalMealPrice });
    // console.log(this.state.extraStatus);
    // console.log(this.state.selectedExtras);
    console.log(totalMealPrice);
  }

  /**
   * @desc runs before component mounts
   * @memberof DocumentView
   * @returns {*} has no return value;
   */
  // componentWillMount() {
  //   this.props.loadMenu().then(() => {
  //     this.setState({
  //       menu: this.props.menu,
  //     });
  //   });
  // }
  render() {
    // const { menu } = this.state;
    const { menu } = this.props;
    const { logout } = this.props;

    // Work on
    // const goeswith = mealExtras => mealExtras.filter(extra => extra.category === 'GoesWith');
    const extras = mealExtras => mealExtras.filter(extra => extra.category === 'OnTop');
    // const meaa = menu.pop();
    console.log(`fdfdcvm,fdf ${menu}`);
    return (
      <div>
        <UserNavBar logout={logout} />
        <div id="container" className="container">
          <h1>Today's Menu</h1>
          {/* <div>
          {menu}
          </div> */}

          <MenuList meals={menu} selected={this.myMeal} />
          {/* get meal from state or something */}
          <MenuModal
            meal={this.state.meal}
            goesWith={this.goesWith}
            extras={this.onTop}
            onQtyChange={this.onQtyChange}
            extraStatus={this.state.extraStatus}
            handleChecked={this.handleChecked}
            totalMealPrice={this.state.totalPrice}
            onSave={this.onSave}
            onClose={this.onClose}
            saving={this.state.saving}
          />
          {/* <MenuModal meal={this.state.meal} /> */}
        </div>
      </div>
    );
  }
}

MenuPage.propTypes = {
  menu: PropTypes.arrayOf(PropTypes.shape).isRequired,
  // menu: PropTypes.shape({
  //   success: PropTypes.bool.isRequired,
  //   message: PropTypes.string.isRequired,
  //   meals: PropTypes.arrayOf(PropTypes.shape).isRequired,
  // })
  //   .isRequired,
  loadMenu: PropTypes.func.isRequired,
  saveOrder: PropTypes.func.isRequired,
  logout: PropTypes.func.isRequired,
  // meal: PropTypes.shape({
  //   id: PropTypes.number.isRequired,
  //   title: PropTypes.string.isRequired,
  //   description: PropTypes.string.isRequired,
  //   price: PropTypes.number.isRequired,
  //   extras: PropTypes.shape.isRequired,
  // })
  //   .isRequired,
};

/**
 * @desc maps dispatch to props;
 * @param {*} dispatch dispatch
 * @returns {*} action to be dispatched
 */
const mapDispatchToProps = dispatch => bindActionCreators(
  {
    loadMenu,
    saveOrder,
    logout,
  }, dispatch);

/**
 * @desc maps state to props;
 * @param {*} state store state
 * @returns {*} store state
 */
const mapStateToProps = (state, ownProps) => {
  // const mealId = ownProps.params.id;
  return ({
    menu: state.menuReducer,
  });
};

const MenuPageWithCSS = CSSModules(MenuPage, styles);
export default connect(mapStateToProps, mapDispatchToProps)(MenuPageWithCSS);
