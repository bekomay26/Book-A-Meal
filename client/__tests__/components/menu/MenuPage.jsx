import React from 'react';
import { shallow, mount, render } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import MenuPage2 from '../../../src/components/menu/MenuPage';
import { MenuPage } from '../../../src/components/menu/MenuPage';
import { getMenuResponse } from '../../__mocks__/mockMenu';
import { getExtrasResponse } from '../../__mocks__/mockExtras';

// import jsdom from 'jsdom';
// const doc = jsdom.jsdom('<!doctype html><html><body></body></html>')
// global.document = doc
// global.window = doc.defaultView
const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const store = mockStore({
  menuReducer: {
    menu: [],
  },
  authReducer: {
    isAuthenticated: true,
    userName: '',
  },
});

const menuArray = [];
const menuArray2 = getMenuResponse.menu.meals;
const extraArray = getExtrasResponse.extras;
function setup() {
  const props = {
    meals: menuArray,
    saveMeal: jest.fn(),
    saveOrder: jest.fn(),
    deleteMeal: jest.fn(),
    loadMeal: jest.fn(),
    loadMenu: jest.fn(),
    extras: [],
  };
  return shallow(<MenuPage {...props} />, reactrouter.get());
}

describe('The Menu Page', () => {
  // it('should render Menu Page component correctly', () => {
  //   const wrapper = setup();
  //   expect(toJson(wrapper)).toMatchSnapshot();
  // });

  it('It renders Drawer, Menu list and the Nav bar', () => {
    // const wrapper = setup();
    const props = {
      menu: menuArray,
      saveOrder: jest.fn(),
      logout: jest.fn(),
      isAuthenticated: true,
      loadMenu: jest.fn(),
    };
    const wrapper = mount(<Provider store={store}><MenuPage {...props} /></Provider>, reactrouter.get());
    expect(wrapper.find('Drawer').length).toEqual(4);
    expect(wrapper.find('MenuList').length).toEqual(1);
    expect(wrapper.find('UserNavBar').length).toEqual(1);
    wrapper.unmount();
    // expect(wrapper.find('Drawer').length).toEqual(1);
  });
});


it('should call the save function', () => {
  const wrapper = setup();
  wrapper.setState({
    saving: null,
    visible: true,
    extraStatus: [
      { isChecked: true, extra: { id: 1 }, qty: 8 },
      { isChecked: false, extra: { id: 2 }, qty: 18 },
      { isChecked: true, extra: { id: 3 }, qty: 28 },
    ],
  });
  const handleSaveSpy = jest.spyOn(wrapper.instance(), 'onSave');
  const redirectSpy = jest.spyOn(wrapper.instance(), 'redirect');
  const event = {
    preventDefault: jest.fn(),
  };
  wrapper.instance().onSave(event);
  expect(handleSaveSpy).toHaveBeenCalled();
  expect(wrapper.state('saving')).toEqual(true);
  expect(wrapper.state('visible')).toEqual(true);
  wrapper.instance().redirect();
  expect(redirectSpy).toHaveBeenCalled();
  expect(wrapper.state('saving')).toEqual(false);
  // expect(spyClose).toHaveBeenCalled();
  // expect(spyRedirect).toHaveBeenCalled();
});

it('should call the updatePredicate function', () => {
  const wrapper = setup();
  window.innerWidth = 961;
  wrapper.setState({ isDesktop: false });
  const handleSaveSpy = jest.spyOn(wrapper.instance(), 'updatePredicate');
  wrapper.instance().updatePredicate();
  expect(handleSaveSpy).toHaveBeenCalled();
  expect(wrapper.state('isDesktop')).toEqual(true);
});

it('should call the myMeal function', () => {
  const wrapper = setup();
  wrapper.setState({ meal: menuArray2[0], basePrice: 100, totalPrice: 250 });
  const handleSaveSpy = jest.spyOn(wrapper.instance(), 'myMeal');
  wrapper.instance().myMeal();
  expect(handleSaveSpy).toHaveBeenCalled();
});

it('should call the goesWith function', () => {
  const wrapper = setup();
  const handleSaveSpy = jest.spyOn(wrapper.instance(), 'goesWith');
  wrapper.instance().goesWith();
  expect(handleSaveSpy).toHaveBeenCalled();
});

it('should call the onTop function', () => {
  const wrapper = setup();
  const handleSaveSpy = jest.spyOn(wrapper.instance(), 'onTop');
  wrapper.instance().onTop();
  expect(handleSaveSpy).toHaveBeenCalled();
});

it('should call the onClose function', () => {
  const wrapper = setup();
  const spy = jest.spyOn(document, 'getElementById');
  spy.mockReturnValue({
    style: {
      display: 'block',
    },
  });
  const handleSaveSpy = jest.spyOn(wrapper.instance(), 'onClose');
  const handleSaveSpy2 = jest.spyOn(wrapper.instance(), 'clearExtrasFields');
  wrapper.instance().onClose();
  expect(handleSaveSpy).toHaveBeenCalled();
  expect(handleSaveSpy2).toHaveBeenCalled();
});

it('should call the quantity changed function', () => {
  const wrapper = setup();
  wrapper.setState({
    meal: menuArray2[0],
    extraStatus: [{ isChecked: true, extra: extraArray[0], key: 1 }, { isChecked: false, extra: extraArray[0], key: 1 }],
  });
  const handleSaveSpy = jest.spyOn(wrapper.instance(), 'onQtyChange');
  const event = {
    preventDefault: jest.fn(),
    target: { value: 1 },
  };
  wrapper.instance().onQtyChange(event, 1, 50);
  expect(handleSaveSpy).toHaveBeenCalled();
});

it('should call the handle checked function', () => {
  const wrapper = setup();
  wrapper.setState({
    meal: menuArray2[0],
    selectedExtras: [],
    extraStatus: [
      {
        isChecked: true, extra: extraArray[0], key: 1, selectedExt: [],
      },
    ],
    totalPrice: 0,
  });
  const handleSaveSpy = jest.spyOn(wrapper.instance(), 'handleChecked');
  const event = {
    preventDefault: jest.fn(),
    target: { value: 1 },
  };
  wrapper.instance().handleChecked(event, 1, extraArray[0], 5);
  expect(handleSaveSpy).toHaveBeenCalled();
});

it('should call the handle unchecked function', () => {
  const wrapper = setup();
  wrapper.setState({
    meal: menuArray2[0],
    selectedExtras: [],
    extraStatus: [
      {
        isChecked: false, extra: extraArray[0], key: 1, selectedExt: [],
      },
    ],
    totalPrice: 0,
  });
  const handleSaveSpy = jest.spyOn(wrapper.instance(), 'handleChecked');
  const event = {
    preventDefault: jest.fn(),
    target: { value: 1 },
  };
  wrapper.instance().handleChecked(event, 1, extraArray[0], 5);
  expect(handleSaveSpy).toHaveBeenCalled();
});

it('should open drawer when a menu meal is clicked', () => {
  // const wrapper = setup();
  const props = {
    menu: menuArray2,
    saveMeal: jest.fn(),
    deleteMeal: jest.fn(),
    loadMeal: jest.fn(),
    loadMenu: jest.fn(),
    extras: [],
  };
  const wrapper = mount(<Provider store={store}><MenuPage {...props} /></Provider>, reactrouter.get());
  wrapper.setState({ meal: menuArray2[0], extraStatus: [{ isChecked: true, extra: extraArray[0], key: 1, selectedExt: [] }] });
  const menuItemArray = wrapper.find('MenuItem');
  expect(menuItemArray.length).toEqual(2);
  menuItemArray.first().simulate('click');
  expect(wrapper.find('Drawer').length).toEqual(4);
  expect(wrapper.find('Button').length).toEqual(1);
  expect(wrapper.find('.ant-drawer-close-x').simulate('click'));
  wrapper.unmount();
  // wrapper.find('Button').simulate('click');
  // const handleSaveSpy = jest.spyOn(wrapper.instance(), 'onSubmit');
  // expect(handleSaveSpy).toHaveBeenCalled(1);
});

it('should show navbar if authenticated', () => {
  store.isAuthenticated = false;
  const props = {
    menu: menuArray2,
    saveMeal: jest.fn(),
    deleteMeal: jest.fn(),
    loadMeal: jest.fn(),
    loadMenu: jest.fn(),
    extras: [],
  };
  const wrapper = mount(<Provider store={store}><MenuPage {...props} /></Provider>, reactrouter.get());
  wrapper.setState({ meal: menuArray2[0], extraStatus: [{ isChecked: true, extra: extraArray[0], key: 1, selectedExt: [] }] });
  const menuItemArray = wrapper.find('MenuItem');
  expect(menuItemArray.length).toEqual(2);
  expect(wrapper.find('#myTopnav').find('.navlink').length).toEqual(13); // should be 4
  expect(wrapper.find('.col-3.navlink.out.show').length).toEqual(3); // should be 0
  wrapper.unmount();
});
// it('should open drawer when a menu meal is clicked', () => {
//   // const wrapper = setup();
//   const props = {
//     menu: menuArray2,
//     saveMeal: jest.fn(),
//     deleteMeal: jest.fn(),
//     loadMeal: jest.fn(),
//     loadMenu: jest.fn(),
//     extras: [],
//   };
//   const wrapper = mount(<Provider store={store}><MenuPage {...props} /></Provider>, reactrouter.get());
//   wrapper.setState({ meal: menuArray2[0], extraStatus: [{ isChecked: true, extra: extraArray[0], key: 1, selectedExt: [] }] });
//   const menuItemArray = wrapper.find('MenuItem');
//   expect(menuItemArray.length).toEqual(2);
//   menuItemArray.first().simulate('click');
//   expect(wrapper.find('Drawer').length).toEqual(4);
//   expect(wrapper.find('Button').length).toEqual(1);
//   expect(wrapper.find('.ant-drawer-close-x').simulate('click'));
//   // wrapper.find('Button').simulate('click');
//   // const handleSaveSpy = jest.spyOn(wrapper.instance(), 'onSubmit');
//   // expect(handleSaveSpy).toHaveBeenCalled(1);
// });
