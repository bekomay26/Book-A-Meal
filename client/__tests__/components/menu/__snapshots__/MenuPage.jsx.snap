// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Menu Page should render Menu Page component correctly 1`] = `
<div>
  <Connect(UserNavBar)
    isAuthenticated={false}
    visibleSideBar={true}
  />
  <div
    className="menu-page-container"
    id="container"
  >
    <h1>
      Today's Menu
    </h1>
    <MenuList
      selected={[Function]}
    />
    <Drawer
      closable={true}
      level={null}
      maskClosable={false}
      onClose={[Function]}
      placement="right"
      prefixCls="ant-drawer"
      style={
        Object {
          "height": "calc(100% - 55px)",
          "overflow": "auto",
          "paddingBottom": 53,
        }
      }
      title="Order Meal"
      visible={false}
      width={400}
    >
      <MenuForm
        extraStatus={Array []}
        extras={[Function]}
        goesWith={[Function]}
        handleChecked={[Function]}
        meal={Object {}}
        onClose={[Function]}
        onQtyChange={[Function]}
        onSave={[Function]}
        saving={false}
        totalMealPrice={0}
      />
    </Drawer>
  </div>
</div>
`;
